### DISCUSSION TITLE
Need help interpreting conflict message

### DISCUSSION THREAD

**Discussion by alaviss** [tree-sitter/tree-sitter#2233] [Q&A] [answered]
I got the following error:

```
Unresolved conflict for symbol sequence:

  'unary_expression_token2'  _simple_expression  •  'import_from_statement_token1'  …

Possible interpretations:

  1:  'unary_expression_token2'  (_command_expression  _simple_expression  •  _simple_expression)                                (precedence: 13, associativity: Left)
  2:  'unary_expression_token2'  (binary_expression  _simple_expression  •  'import_from_statement_token1'  _simple_expression)  (precedence: 6, associativity: Left)
  3:  (unary_expression  'unary_expression_token2'  _simple_expression)  •  'import_from_statement_token1'  …                    (precedence: 12, associativity: Left)

Possible resolutions:

  1:  Specify a higher precedence in `_command_expression` and `binary_expression` than in the other rules.
  2:  Specify a higher precedence in `unary_expression` than in the other rules.
  3:  Add a conflict for these rules: `_command_expression`, `unary_expression`, `binary_expression`
```

From the looks of it, the precedence has already been laid out in a strictly increasing order, so why couldn't tree-sitter resolves according to that precedence?

**Comment by ahlinc** [comment, marked as answer]
It looks like the grammar rules defined in a way that Tree-sitter can't make a choice between them by doing a one step terminal token lookahead. If it's not possible or not acceptable to split rules to more granular with reused common parts then try to assign rule conflicts like suggested by Tree-sitter, this will enable lookaheads for more that one token to distinguish between rules.

**Reply by alaviss** [reply]
Then I suppose it's facing the choice between `'import_from_statement_token1'` and `_simple_expression <- identifier`.

However, my language does not permit `import_from_statement_token1` to be resolved as an identifier (ie. globally reserved keyword), is there anyway I can have tree-sitter be aware of this and remove that proposition altogether?

**Reply by ahlinc** [reply]
Sorry but without a grammar example it's not easy to understand a full context of your issue to advice something more.

**Comment by alaviss** [comment]
So here's a minimal grammar I extracted:

```js
module.exports = grammar({
  name: "example",

  word: ($) => $.identifier,
  precedences: ($) => [
    [
      $.command_expression,
      "unary",
      "bin4",
      "bin2",
      $._command_expression_arguments,
    ],
  ],
  rules: {
    source: ($) => repeat(seq($._expression, "\n")),
    _expression: ($) =>
      choice(
        $.binary_expression,
        $.command_expression,
        $.unary_expression,
        $.identifier
      ),

    binary_expression: ($) =>
      choice(
        prec.left("bin4", seq($._expression, "not", $._expression)),
        prec.left("bin2", seq($._expression, "and", $._expression))
      ),
    unary_expression: ($) =>
      prec.left("unary", seq(choice("not", "and"), $._expression)),
    command_expression: ($) =>
      seq($._expression, $._command_expression_arguments),
    _command_expression_arguments: ($) => $._expression,

    identifier: ($) => /[a-zA-Z\x80-\xff](_?[a-zA-Z0-9\x80-\xff])*/,
  },
});
```

I'd like to parse this

```
not a not b
not a and not b
a not a b
a not b b not a
```

as

```
(source
  (binary_expression
    (unary_expression
      (identifier))
    (identifier))
  (binary_expression
    (unary_expression
      (identifier))
    (unary_expression
      (identifier)))
  (binary_expression
    (identifier)
    (command_expression
      (identifier)
      (identifier)))
  (binary_expression
    (identifier)
    (command_expression
      (identifier)
      (binary_expression
        (identifier)
        (identifier)))))
```

**Reply by alaviss** [reply]
Now that I wrote this out, I can see the issue where `not a not b` can be parsed as `not(a(not(b)))`

I'm a bit loss on how to structure the precedence structure to make the left hand side of `command_expression` binds stronger than all expressions but the entire thing as a whole is weaker than `binary_expression`

**Comment by alaviss** [comment]
So I'm getting some other confusing messages:

```
Unresolved conflict for symbol sequence:

  _layout_start  _line  _statement  •  ';'  …

Possible interpretations:

  1:  _layout_start  _line  (_block_line_statement_list  _statement  •  _block_line_statement_list_repeat1)
  2:  _layout_start  _line  (_statement  _statement)  •  ';'  …

Possible resolutions:

  1:  Specify a higher precedence in `_block_line_statement_list` than in the other rules.
  2:  Specify a higher precedence in `_statement` than in the other rules.
  3:  Specify a left or right associativity in `_statement`
  4:  Add a conflict for these rules: `_block_line_statement_list`, `_statement`
```

Here, my `_block_line_statement_list` is simply `seq($._statement, repeat(seq(";", $._statement)))`.

As far as I can tell, there's nothing in the grammar itself that should be ambiguous here, if you see a `;` then it's a `_block_line_statement_list_repeat1`.

I can't reproduce this on a mini grammar, so I don't know how to debug this other than trimming my current grammar until it stop throwing errors, which is very time consuming.

Is there any way to have tree-sitter explains itself fully here?

**Reply by alaviss** [reply]
Never mind, I made a loop and I didn't notice it `_statement <- _statement`
